jxl_cli/src/bin/jxlinspect.rs

```diff
@@ -60,74 +63,133 @@ fn parse_jxl_codestream(data: &[u8], verbose: bool) -> Result<(), jxl::error::Er
+    );
+    let alpha_info = match file_header
+        .image_metadata
+        .extra_channel_info
+        .iter()
+        .any(|info| info.alpha_associated())
+    {
+        true => "+Alpha",
+        false => "",
+    };
+    let image_or_animation = match file_header.image_metadata.animation {
+        None => "Image",
+        Some(_) => "Animation",
+    };
+    print!(
+        "JPEG XL {}, {}x{}, {}, {}-bit {}{}",
+        image_or_animation,
+        file_header.size.xsize(),
+        file_header.size.ysize(),
+        how_lossy,
+        file_header.image_metadata.bit_depth.bits_per_sample(),
+        color_space,
+        alpha_info,
+    );
+    if file_header
+        .image_metadata
+        .bit_depth
+        .exponent_bits_per_sample()
+        != 0
+    {
         print!(
-            "JPEG XL {}, {}x{}, {}, {}-bit {}{}",
-            image_or_animation,
-            file_header.size.xsize(),
-            file_header.size.ysize(),
-            how_lossy,
-            file_header.image_metadata.bit_depth.bits_per_sample(),
-            color_space,
-            alpha_info,
+            "float ({} exponent bits)",
+            file_header
+                .image_metadata
+                .bit_depth
+                .exponent_bits_per_sample()
         );
-        if file_header
-            .image_metadata
-            .bit_depth
-            .exponent_bits_per_sample()
-            != 0
-        {
-            print!(
-                "float ({} exponent bits)",
-                file_header
-                    .image_metadata
-                    .bit_depth
-                    .exponent_bits_per_sample()
-            );
-        }
-        println!();
+    }
+    println!();
+    if file_header.image_metadata.color_encoding.want_icc {
+        println!("with ICC profile")
+    } else {
+        print_color_encoding(&file_header.image_metadata.color_encoding);
+    }
+    if verbose {
+        // Verbose output: Use Debug trait to print the FileHeaders
+        println!("{:#?}", file_header);
+        // TODO(firsching): consider printing more of less information for ICC
+        // for verbose/non-verbose cases
         if file_header.image_metadata.color_encoding.want_icc {
-            println!("with ICC profile")
-        } else {
-            print_color_encoding(&file_header.image_metadata.color_encoding);
+            let icc_data = read_icc(&mut br)?;
+            println!("ICC profile length: {} bytes", icc_data.len());
         }
-        return Ok(());
     }
+    //TODO(firsching): handle frames which are blended together, also within animations.
+    if file_header.image_metadata.animation.is_some() {
+        let mut total_duration = 0.0f64;
+        let animation = file_header
+            .image_metadata
+            .animation
+            .as_ref()
+            .expect("This should never fail, it was just check in the if condition above");
+        let mut not_is_last = true;
+        while not_is_last {
```

Comment:
Any thoughts about making this a `fn for_each_frame<F: FnMut(FrameHeader, Toc, ...) -> Result<bool>>(br: BitReader, f: F) -> Result<()>` or similar? (with the function returning Err for errors, Ok(true) for "continue to next frame", and Ok(false) for "exit")

Reply:
General Rust question: What are the benefits of a function with a callback as opposed to an iterator?

Reply:
Iterators are harder to write :-)

I would also be OK with an iterator, fwiw.

Reply:
will do in a follow-up, added TODO